Closes #8423: Allow assigning Service to FHRP Group, in addition to Device and VirtualMachine

jnovinger opened the pull request:
### Closes: #8423

- Allow assigning Service to FHRP Group, in addition to Device and VirtualMachine
- Changes relation type for all three to a single GenericForeignKey that is required
  - Has the same semantics as before, that were enforced in `Service.clean()`, now enforced by the GFK and in model forms
- Patches up REST API and GraphQL representations
- Fixes up `ServiceImportForm` to work with the new `Service.parent` GFK

jnovinger committed:
Allow assigning Service to FHRP Group, in addition to Device and VirtualMachine

jnovinger committed:
Finish up add FHRP Group assignment to Service

- fixes up ServiceSerializer to support write operations
- fixes up GraphQL components: ServiceType and ServiceFilter
- fixes broken tests
- cleans up lint issues

jnovinger committed:
Remove unnecessary limit_choices_to arg in new Service.parent_object_type field

Re: #17166

arthanson reviewed (changes requested):
Also - not sure if we want to change from tabbed interface for selection of assignment to dropdown - see discussion in maintainers channel chat.

jnovinger commented on netbox/ipam/filtersets.py:
```diff
@@ -1137,25 +1137,35 @@ def search(self, queryset, name, value):
-        queryset=Device.objects.all(),
-        to_field_name='name',
-        label=_('Device (name)'),
+    virtual_machine = MultiValueCharFilter(
```
🤦

jnovinger replied:
> I think? we can get rid of the MultiValueCharFitler fields for these with just a single assignment_type ContentTypeFilter - see VLANGroupFilterSet scope_type.

Actually, I would argue that the scope filters on VLAN Groups are buggy. For instance, using the demo data import, I can select more than one site to filter on--say 'JBB Branch 104' and 'JBB Branch 109'. However, it's only actually ever using the second one of those (guessing the last one to appear in the query params) for any filtering (see screenshot).

![image](https://github.com/user-attachments/assets/b9412364-76a4-4c70-b6c2-9b7a354be3f2)

jnovinger replied:
> Need to add fhrpgroup to the filter form

Fixed

jnovinger committed:
Address PR feedback, add FHRP Group to ServiceFilterForm

jnovinger committed:
Address PR feedback, clean up migrations

jnovinger committed:
Address PR feedback, add Services to FHRP Group related objects

jnovinger committed:
Address PR feedback, add ServiceFilterSet test for fhrpgroup

jnovinger committed:
Fix Service detail page breadcrumb links

jnovinger committed:
Address PR feedback, switch parent GFK to dropdown on forms

a la ScopedForm

jnovinger committed:
Fix broken tests from changes to ServiceForm/ServiceCreateForm

jnovinger committed:
Fixup multi-if to use match statement

jnovinger commented:
> FHRP Group details view should show the related services, either as related objects or a sub-table.

Fixed

arthanson commented:
@jnovinger test failing, not sure why: netbox/ipam/views.py:1435:9: SyntaxError: Cannot use `match` statement on Python 3.9 (syntax was added in Python 3.10)

jeremystretch commented:
@arthanson see #19112

jnovinger commented:
> @jnovinger test failing, not sure why: netbox/ipam/views.py:1435:9: SyntaxError: Cannot use `match` statement on Python 3.9 (syntax was added in Python 3.10)

Ah, this is actually a linting thing. I guess this is what the PR from Jeremy re: ruff and 3.10 python target is all about.

jnovinger commented:
Just realized I also missed updating the CSV import form for Services. Working on that now.

jnovinger committed:
Fixes up ServiceImportForm to work with new parent field

jnovinger commented on netbox/ipam/forms/bulk_import.py:
```diff
@@ -572,15 +570,43 @@ class ServiceImportForm(NetBoxModelImportForm):
                 )
 
-        return self.cleaned_data['ipaddresses']
+        return self.cleaned_data
```
All of this logic got moved up to `clean()` from `clean_ipaddresses()` because it needs access to a resolved value for the `parent` field. I tried doing that within the existing `clean_ipaddresses()`, but it was a mess and didn't feel reliable.

jnovinger commented on netbox/ipam/forms/bulk_import.py:
```diff
@@ -572,15 +570,43 @@ class ServiceImportForm(NetBoxModelImportForm):
+        elif not parent and (parent_id := self.cleaned_data.get('parent_object_id')):
+            ct = self.cleaned_data.get('parent_object_type')
+            parent = ct.model_class().objects.filter(id=parent_id).first()
+            self.cleaned_data['parent'] = parent
```
`ServiceImportForm` supports either using the (`parent_object_type`, `parent`) field pair or the (`parent_object_type`, `parent_object_id`) field pair. This `if` and `elif` handle making sure they are both resolved when the other is used.

I don't feel fantastic about this approach. If there's pushback I'm happy to switch it to only support the (`parent_object_type`, `parent`) field pair.

jnovinger committed:
Raise ValidationError if parent object form data is incomplete

arthanson commented on netbox/ipam/forms/model_forms.py:
```diff
@@ -21,7 +21,7 @@
 from utilities.forms.utils import get_field_value
```
If you set parent type without the associated device, VM, FHRP Group and try to save there is no error given, should be a "Cannot set parent type without an associated object" message

jnovinger replied:
Phew, this only took forever to figure out. After a fair amount of debugging compiled .ts, reading bootstrap docs, and finally understanding how those in-page "This field is required." messages get displayed, I discovered this is because of #18916.

The **tl;dr** is that the `APISelector`'s template changes the DOM tree structure just enough to break the way Bootstrap display's invalid feedback for fields. See [my comment on the issue](https://github.com/netbox-community/netbox/issues/18916#issuecomment-2791189429).

jnovinger replied:
I think the work-around for this one is to just use Django's built in `Select` widget, since with the `HTMXSelect` widget on the `parent_object_type` form field, the form is being rendered and a queryset attached to the `parent` form field select, we don't necessarily need the `APISelect` widget functionality.

The downside is that means no object selector modal for the `parent` form field for the time being.

Alternatively, we solve #18916 real quick.

jnovinger replied:
After discussing this in the maintainers call, we decided to move forward with this work as is re: the lack of validation messaging working in the `DynamicModelChoiceField`. #18916 will be a follow up in a future patch release.

arthanson reviewed (changes requested):
Just a couple minor issues - also needs docs in docs/models/ipam/service.md add in the parent field, can look at scope on vlan group for template.

jnovinger committed:
Update docs to reflect change of Service.parent in v4.3

jnovinger committed:
Address PR feedback, ensure that parent GFK underlying fields are not nullable

This changes an existing (within the branch) migration, so if you're
testing this you'll need to back up migrations and re-run.

jnovinger committed:
Fix migration conflict after rebase on feature branch

jnovinger commented:
@arthanson , please note that the migrations have changed, for a couple of reasons:
1. there was a conflict with a migration from another branch that got merged ahead of this one
2. I had forgotten to enforce that `Service.parent` could not be null on the model itself.

You'll need at least wind back the previous IPAM `0078`, `0079`, and `0080` migrations before trying to run these.

jnovinger committed:
Fix up migration dependency after rebase
